phantasm-at-rest/src/main/java/com/chiralbehaviors/CoRE/phantasm/resources/TransactionalResource.java

```diff
@@ -96,46 +94,67 @@ protected Model getNewModel() {
         return new ModelImpl(emf);
     }
 
-    protected UUID insert(final Ruleform ruleform) {
-        return perform((Model model) -> Util.smartMerge(model.getEntityManager(),
-                                                        ruleform,
-                                                        new HashMap<>())
-                                            .getId());
-    }
-
-    protected <T> T perform(Function<Model, T> txn) throws WebApplicationException {
+    protected <T> T perform(AuthorizedPrincipal principal,
+                            Function<Model, T> txn) throws WebApplicationException {
         Model model = new ModelImpl(emf);
         EntityManager em = model.getEntityManager();
         em.getTransaction()
           .begin();
+        if (principal == null) {
+            principal = new AuthorizedPrincipal(model.getKernel()
+                                                     .getUnauthenticatedAgency());
+        } else {
+            principal = principal.merge(em);
+        }
         try {
-            T value = txn.apply(model);
-            em.getTransaction()
-              .commit();
-            return value;
-        } catch (Throwable e) {
+            return model.executeAs(principal, () -> {
+                try {
+                    T value = txn.apply(model);
+                    em.getTransaction()
+                      .commit();
+                    return value;
+                } catch (Throwable e) {
+                    log.error("error applying transaction", e);
+                    throw new WebApplicationException(e, 500);
+                } finally {
+                    em.close();
+                }
+            });
+        } catch (Exception e) {
             log.error("error applying transaction", e);
             throw new WebApplicationException(e, 500);
-        } finally {
-            em.close();
         }
     }
 
-    protected <T> T readOnly(Function<Model, T> txn) throws WebApplicationException {
+    protected <T> T readOnly(AuthorizedPrincipal principal,
+                             Function<Model, T> txn) throws WebApplicationException {
         Model model = new ModelImpl(emf);
         EntityManager em = model.getEntityManager();
         em.getTransaction()
           .begin();
         em.getTransaction()
           .setRollbackOnly();
+        if (principal == null) {
```

Comment:
same thing here.  probably should refactor this to pull out the common code between perform and readonly